### PR TITLE
Merging new DICOM IO connector changes with Beam Repository

### PR DESCRIPTION
**Pull request by Aliraza-N**
Changes:
- new DICOM IO connector is created that allows DICOM API to be used by Beam
- New methods in existing Healthcare API client class

### CODE REVIEW

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java
Comment (Aliraza-N) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java:16]: Need to revert this

Comment (dranderson1117) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java]: I would leave all the imports individually defined to allow other contributors a better idea of what is being used.

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java
Comment (dranderson1117) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java]: Leave the individual imports explicitly defined for better understanding of what is being used.

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java
Comment (dranderson1117) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java:214]: add javadoc for createDicomStore

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (lastomato) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:32]: You can wrap this block in <pre></pre>, also please break into multiple lines and add semicolons.

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (lastomato) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:158]: Catching all exceptions is a bad practice: see https://stackoverflow.com/questions/2416316/why-is-the-catchexception-almost-always-a-bad-idea

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java
Comment (lastomato) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java:203]: nit: we should probably create a utility class for handling path parsing, since it can be useful for other modalities as well.

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java
Comment (lastomato) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java:229]: Is Nullable the right annotation?

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java
Comment (lastomato) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HttpHealthcareApiClient.java:259]: Is it possible to import the HttpResponse class? Or you can just do:

```java
return request.executeUnparsed().parseAsString();
```

**Review by lastomato — COMMENTED** [review]
Please add tests for at least the DicomIO.

**Comment by Aliraza-N** [comment]
@lastomato How do these changes look?

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/WebPathParser.java
Comment (lastomato) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/WebPathParser.java:14]: Please add unit tests for this class.

**Review by lastomato — COMMENTED** [review]
LGTM. Just a qq before I approve, what is our plan to submit the DICOM IO to the beam SDK?

**Comment by Aliraza-N** [comment]
There is an open PR in that repo for the same changes. I'll go back and update these new changes there as well.

**Comment by Aliraza-N** [comment]
@lastomato Ready for review please

**Comment by lastomato** [comment]
> There is an open PR in that repo for the same changes. I'll go back and update these new changes there as well.

Please make sure to add integration tests to that PR.

**Review by lastomato — APPROVED** [review]
Please do not submit this PR directly, rather ask your host to approve the internal version of this PR.

**Comment by Aliraza-N** [comment]
> > There is an open PR in that repo for the same changes. I'll go back and update these new changes there as well.
> 
> Please make sure to add integration tests to that PR.

Yes, there are already IT in that repo for DICOM IO

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:26]: You can merge the first two lines here into something like: 

"The study-level metadata for a dicom instance can be read with {@link ReadStudyMetadata} using the study path as a string"

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:110]: "Gets the read metadata of the resource"?

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:146]: can just call it newHealthcareClient?

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HealthcareApiClient.java:226]: store if of the new dicom store?

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/WebPathParser.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/WebPathParser.java:28]: You want to check the length of the Array here before accessing the indexes to be sure we never run into ArrayIndexOutOfBounds exception..

**Review by poojavenkatram — COMMENTED** [review]
Hi Aliraza,

Looks good overall a few small comments and then we are good to go.

Best
Pooja

**Review by poojavenkatram — APPROVED** [review]
Thanks Aliraza - LGTM.

**Comment by dranderson1117** [comment]
Nice job Aliraza - LGTM

**Review by dranderson1117 — COMMENTED** [review]
closed PR by accident...reopening

**Comment by dranderson1117** [comment]
reopening PR

**Review by dranderson1117 — APPROVED** [review]
LGTM

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:38]: this is too long

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:39]: this is too long...

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:66]: this is too long..

File: src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java
Comment (poojavenkatram) [inline src/main/java/org/apache/beam/sdk/io/gcp/healthcare/DicomIO.java:153]: this is too long too

File: src/test/java/org/apache/beam/sdk/io/gcp/healthcare/WebPathParserTest.java
Comment (poojavenkatram) [inline src/test/java/org/apache/beam/sdk/io/gcp/healthcare/WebPathParserTest.java:30]: considering you are splitting this anyway wanna split it as <100chars in one line and the rest on the next?

**Pull request closed without merging**
